src/concurrency/timestamp_ordering_transaction_manager.cpp

```diff
@@ -84,14 +85,55 @@ void TimestampOrderingTransactionManager::InitTupleReserved(
   *(cid_t *)(reserved_area + LAST_READER_OFFSET) = 0;
 }
 
-TimestampOrderingTransactionManager &
-TimestampOrderingTransactionManager::GetInstance() {
+Transaction *TimestampOrderingTransactionManager::BeginTransaction() {
+  txn_id_t txn_id = GetNextTransactionId();
+  cid_t begin_cid = GetNextCommitId();
+  Transaction *txn = new Transaction(txn_id, begin_cid);
+
+  auto eid = EpochManagerFactory::GetInstance().EnterEpoch(begin_cid);
+  txn->SetEpochId(eid);
+
+  if (FLAGS_stats_mode != STATS_TYPE_INVALID) {
+    stats::BackendStatsContext::GetInstance()
+        .GetTxnLatencyMetric()
+        .StartTimer();
+  }
+
+  return txn;
+}
+
+void TimestampOrderingTransactionManager::EndTransaction(Transaction *current_txn) {
+  EpochManagerFactory::GetInstance().ExitEpoch(current_txn->GetEpochId());
+
+  if (current_txn->GetResult() == RESULT_SUCCESS) {
+    gc::GCManagerFactory::GetInstance().
+        RecycleTransaction(current_txn->GetGCSetRef(), current_txn->GetBeginCommitId(), GC_SET_TYPE_COMMITTED);
+  } else {
+    gc::GCManagerFactory::GetInstance().
+        RecycleTransaction(current_txn->GetGCSetRef(), GetNextCommitId(), GC_SET_TYPE_ABORTED);
+  }
+
+  delete current_txn;
+  current_txn = nullptr;
+
+  if (FLAGS_stats_mode != STATS_TYPE_INVALID) {
+    stats::BackendStatsContext::GetInstance()
+        .GetTxnLatencyMetric()
+        .RecordLatency();
+  }
+}
+
+
+TimestampOrderingTransactionManager &TimestampOrderingTransactionManager::GetInstance() {
   static TimestampOrderingTransactionManager txn_manager;
   return txn_manager;
 }
 
 bool TimestampOrderingTransactionManager::IsOccupied(
```

Comment:
This function is similar to the IsVisible() method. May be we can refactor it to reduce duplicated code.